src/value/de.rs

```diff
@@ -41,22 +41,19 @@ impl<'de> Deserialize<'de> for Value {
             fn visit_i64<E>(self, i: i64) -> Result<Value, E>
                 where E: SError,
             {
-                Ok(Value::I64(i))
+                Ok(Value::Number(Number::from(i)))
             }
 
             fn visit_u64<E>(self, u: u64) -> Result<Value, E>
                 where E: SError,
             {
-                match NumCast::from(u) {
-                    Some(i) => Ok(Value::I64(i)),
-                    None => Ok(Value::String(u.to_string())),
-                }
+                Ok(Value::Number(Number::from(u)))
             }
 
             fn visit_f64<E>(self, f: f64) -> Result<Value, E>
                 where E: SError,
             {
-                Ok(Value::F64(f))
+                Ok(Value::Number(Number::from_f64(f).expect("wrapping an existing float")))
```

Comment:
Ok, cool. So I'm thinking two directions. One

```
enum N {
    PosInt(u64),
    /// Always less than zero.
    NegInt(i64),
    /// Always finite.
    Float(f64),
    NaN,
    InfinityPos,
    InfinityNeg
}
```

but that feels like its mixes up float concerns and non float concerns.

Instead add an enum to wrap the float values like: 

```
enum F {
    PosInfinity,
    NegInfinity,
    Finite(f64),
    NaN,
}
```

so that `N` becomes

```
enum N {
    PosInt(u64),
    NegInt(i64),
    Float(F),
}
```

Does that jive with you?